File: src/main/scala/com/microsoft/azure/cosmosdb/kafka/connect/source/CosmosDBSourceConnector.scala
package com.microsoft.azure.cosmosdb.kafka.connect.source

import java.util

import com.microsoft.azure.cosmosdb._
import com.microsoft.azure.cosmosdb.kafka.connect.common.ErrorHandling.ErrorHandler

import scala.collection.JavaConversions._
import com.microsoft.azure.cosmosdb.{ConnectionPolicy, ConsistencyLevel}
import com.microsoft.azure.cosmosdb.kafka.connect.{CosmosDBClientSettings, CosmosDBProvider}
import com.microsoft.azure.cosmosdb.kafka.connect.config.{ConnectorConfig, CosmosDBConfig, CosmosDBConfigConstants}
import com.typesafe.scalalogging.{LazyLogging, StrictLogging}
import org.apache.kafka.common.config.ConfigDef
import org.apache.kafka.connect.connector.Task
import org.apache.kafka.connect.source.SourceConnector
import org.apache.kafka.connect.util.ConnectorUtils
import scala.util.{Failure, Success, Try}
import scala.collection.JavaConverters._

class CosmosDBSourceConnector extends SourceConnector with StrictLogging with ErrorHandler {

  private var configProps: util.Map[String, String] = _
  private var numWorkers: Int = 0

  override def version(): String = getClass.getPackage.getImplementationVersion

  override def start(props: util.Map[String, String]): Unit = {
    logger.info("Starting CosmosDBSourceConnector")
    configProps = props
  }

  override def taskClass(): Class[_ <: Task] = classOf[CosmosDBSourceTask]

  override def taskConfigs(maxTasks: Int): util.List[util.Map[String, String]] = {
    initializeErrorHandler(2)
    try {
      val config: CosmosDBConfig = CosmosDBConfig(ConnectorConfig.sourceConfigDef, configProps)
      val database: String = config.getString(CosmosDBConfigConstants.DATABASE_CONFIG)
      val collection: String = config.getString(CosmosDBConfigConstants.COLLECTION_CONFIG)
      val settings: CosmosDBClientSettings = CosmosDBClientSettings(
        config.getString(CosmosDBConfigConstants.CONNECTION_ENDPOINT_CONFIG),
        config.getPassword(CosmosDBConfigConstants.CONNECTION_MASTERKEY_CONFIG).value(),
        database,
        collection,
        ConnectionPolicy.GetDefault(),
        ConsistencyLevel.Session
      )
      logger.debug("Settings for Cosmos Db connection: ", settings)

      val client = CosmosDBProvider.getClient(settings)

      val collectionLink = CosmosDBProvider.getCollectionLink(database, collection)
      val changeFeedObservable = client.readPartitionKeyRanges(collectionLink, null)
      var results = List[PartitionKeyRange]()
      changeFeedObservable.toBlocking().forEach(x => results = results ++ x.getResults())
      val numberOfPartitions = results.map(p => p.getId)
      numWorkers = Math.min(numberOfPartitions.size(), maxTasks)
      logger.info(s"Setting task configurations for $numWorkers workers.")
      val groups = ConnectorUtils.groupPartitions(numberOfPartitions, maxTasks)
      groups
        .withFilter(g => g.nonEmpty)
        .map { g =>
          val taskConfigs = new java.util.HashMap[String, String](this.configProps)
          taskConfigs.put(CosmosDBConfigConstants.ASSIGNED_PARTITIONS, g.mkString(","))
          taskConfigs
        }
    }
    catch {
      case f: Throwable =>
        logger.error(s"Couldn't initialize CosmosDb with settings: ${f.getMessage}", f)
        HandleError(Failure(f))
        return null
    }
  }

  override def config(): ConfigDef = ConnectorConfig.sourceConfigDef

  override def stop(): Unit = {
    logger.info("Stopping CosmosDBSourceConnector")
  }

  def getNumberOfWorkers(): Int = numWorkers

}

File: src/main/java/com/microsoft/azure/cosmosdb/kafka/connect/source/CosmosDBSourceConnector.java
package com.microsoft.azure.cosmosdb.kafka.connect.source;

import org.apache.commons.lang3.StringUtils;
import org.apache.kafka.common.config.ConfigDef;
import org.apache.kafka.connect.connector.Task;
import org.apache.kafka.connect.source.SourceConnector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

/**
 * The CosmosDB Source Connector
 */
public class CosmosDBSourceConnector extends SourceConnector {
    private static final Logger logger = LoggerFactory.getLogger(CosmosDBSourceConnector.class);
    private SourceSettings settings = null;

    @Override
    public void start(Map<String, String> sourceConectorSetttings) {
        logger.info("Starting the Source Connector");
        this.settings = new SourceSettings();
        this.settings.populate(sourceConectorSetttings);
    }

    @Override
    public Class<? extends Task> taskClass() {
        return CosmosDBSourceTask.class;
    }


    @Override
    public List<Map<String, String>> taskConfigs(int maxTasks) {
        logger.info("Creating the task Configs");
        String[] containerList = StringUtils.split(settings.getContainerList(),",");
        List<Map<String, String>> taskConfigs = new ArrayList<>(maxTasks);

        if(containerList.length == 0) {
            logger.debug("Container list is not specified");
            return taskConfigs;
        }
        for (int i = 0; i< maxTasks; i++) {
            // Equally distribute workers by assigning workers to containers in round robin fashion.
            this.settings.setAssignedContainer(containerList[i % containerList.length]);
            this.settings.setWorkerName("worker" + i);
            Map<String, String> taskConfig = this.settings.asMap();
            taskConfigs.add(taskConfig);
        }

        return taskConfigs;
    }


    @Override
    public void stop() {
        logger.info("Stopping CosmosDBSourceConnector");
    }

    @Override
    public ConfigDef config() {
        ConfigDef configDef = new ConfigDef();
        new SourceSettings().getAllSettings().stream().forEachOrdered(setting -> setting.toConfigDef(configDef));
        return configDef;
    }

    @Override
    public String version() {
        return this.getClass().getPackage().getImplementationVersion();
    }
}
